extends BasePuppetScript
## SCP-023 puppet script.
## It is like a delayed timeb0mb with warning.
## It appears in late round, you need to come to repair eyes. Won't do it - catch gameover.
## 
class_name Scp023PuppetScript

var eye_glow_strength: float = 0.25

@export var glow_enabled: bool = true
@onready var timer: Timer = $Timer

func on_start():
	if Settings.setting_res.zen_mode:
		glow_enabled = false
	if glow_enabled:
		timer.wait_time = rng.randf_range(155, 192)
		timer.start()

func _physics_process(delta: float) -> void:
	match state:
		States.IDLE:
			call("set_state", "idle")
		States.WALKING:
			call("set_state", "walk")
	$rig_001_deform/Skeleton3D/Plane.mesh.surface_get_material(2).set_shader_parameter("emission_strength", eye_glow_strength)
	if !timer.is_stopped():
		eye_glow_strength = lerpf(0.25, 2.0, (timer.wait_time - timer.time_left) / timer.wait_time )
		if eye_glow_strength > 1.75:
			if !get_tree().root.get_node("Game/FoundationTask").has_task("task_023_emergency"):
				get_tree().root.get_node("Game/FoundationTask").trigger_event(2, load("res://Scripts/TaskSystem/Tasks/Scp023EmergencyTask.tres"))


## Animation state
func set_state(anim_name: String) -> void:
	# if animation is the same, do nothing, else play new animation
	if $AnimationPlayer.current_animation == anim_name:
		return
	$AnimationPlayer.play(anim_name, 0.3)


func _on_timer_timeout() -> void:
	get_tree().root.get_node("Game").finish_game(false, "GAME_OVER_4")

func special_action():
	if glow_enabled:
			eye_glow_strength = 0.25
			$Timer.stop()
			if get_tree().root.get_node("Game/FoundationTask").has_task("task_023_emergency"):
				get_tree().root.get_node("Game/FoundationTask").get_tree().root.get_node("Game/FoundationTask").trigger_event(0)
